File: Arithmetic/0022.GenerateParentheses/solution.php
<?php

class Solution {

    private $ans = [];

    function generateParenthesis($n) {
        $this->make(0, 0 , $n, '');
        return $this->ans;
    }

    function make($left, $right, $max, $str) {
        if($right == $max){
            $this->ans[] = $str;
        }else{
            if($left < $max){
                $this->make( $left + 1, $right, $max, $str . '(');
            }

            if($right < $left){
                $this->make($left, $right + 1, $max, $str . ')');
            }
        }
    }

}

$solution = new Solution;
$ans = $solution->generateParenthesis(3);
var_dump($ans);

File: Arithmetic/0022.GenerateParentheses/solution.go
package main

import(
	"fmt"
)

func main(){
	num := 3
	ans := generateParenthesis(num)
	fmt.Println(ans)
}

var ans []string = []string{}
func generateParenthesis(n int) []string {
	make(0, 0, n, "")
	return ans
}

func make(left, right, max int, str string){
	if right == max {
		ans = append(ans, str)
	}
	if(left < max){
		make(left + 1, right, max, str + "(")
	}
	if(right < left){
		make(left, right + 1, max, str + ")")
	}
}